web_ui/authentication.go

```diff
@@ -0,0 +1,297 @@
+		IssuedAt(now).
+		Expiration(now.Add(30 * time.Minute)).
+		NotBefore(now).
+		Subject(user).
+		Build()
+	if err != nil {
+		ctx.JSON(500, gin.H{"error": "Failed to build token"})
+		return
+	}
+	log.Debugf("Type of *key: %T\n", key)
+	var raw ecdsa.PrivateKey
+	if err = key.Raw(&raw); err != nil {
+		ctx.JSON(500, gin.H{"error": "Unable to sign login cookie"})
+		return
+	}
+	signed, err := jwt.Sign(tok, jwt.WithKey(jwa.ES256, raw))
+	if err != nil {
+		log.Errorln("Failure when signing the login cookie:", err)
+		ctx.JSON(500, gin.H{"error": "Unable to sign login cookie"})
+		return
+	}
+
+	ctx.SetCookie("login", string(signed), 30*60, "/api/v1.0",
+		ctx.Request.URL.Host, true, true)
+	// Explicitly set Cookie for /metrics endpoint as they are in different paths
+	ctx.SetCookie("login", string(signed), 30*60, "/metrics",
+		ctx.Request.URL.Host, true, true)
+	ctx.SetSameSite(http.SameSiteStrictMode)
+}
+
+// Check if user is authenticated by checking if the "login" cookie is present and set the user identity to ctx
+func authHandler(ctx *gin.Context) {
+	user, err := getUser(ctx)
+	if err != nil || user == "" {
+		log.Errorln("Invalid user cookie or unable to parse user cookie:", err)
+		ctx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Authentication required to perform this operation"})
+	} else {
+		ctx.Set("User", user)
+		ctx.Next()
+	}
+}
+
+// Handle regular username/password based login
+func loginHandler(ctx *gin.Context) {
+	db := authDB.Load()
+	if db == nil {
+		newPath := path.Join(ctx.Request.URL.Path, "..", "initLogin")
+		initUrl := ctx.Request.URL
+		initUrl.Path = newPath
+		ctx.Redirect(307, initUrl.String())
+		return
+	}
+
+	login := Login{}
+	if ctx.ShouldBind(&login) != nil {
+		ctx.JSON(400, gin.H{"error": "Missing user/password in form data"})
+		return
+	}
+	if !db.Match(login.User, login.Password) {
+		ctx.JSON(401, gin.H{"error": "Login failed"})
+		return
+	}
+
+	setLoginCookie(ctx, login.User)
+	ctx.JSON(200, gin.H{"msg": "Success"})
+}
+
+// Handle initial code-based login for admin
+func initLoginHandler(ctx *gin.Context) {
+	db := authDB.Load()
+	if db != nil {
+		ctx.JSON(400, gin.H{"error": "Authentication is already initialized"})
+		return
+	}
+	curCode := currentCode.Load()
+	if curCode == nil {
+		ctx.JSON(400, gin.H{"error": "Code-based login is not available"})
+		return
+	}
+	prevCode := previousCode.Load()
+
+	code := InitLogin{}
+	if ctx.ShouldBind(&code) != nil {
+		ctx.JSON(400, gin.H{"error": "Login code not provided"})
+		return
+	}
+
+	if code.Code != *curCode && (prevCode == nil || code.Code != *prevCode) {
+		ctx.JSON(401, gin.H{"error": "Invalid login code"})
+		return
+	}
+
+	setLoginCookie(ctx, "admin")
+}
+
+// Handle reset password
+func resetLoginHandler(ctx *gin.Context) {
+	passwordReset := PasswordReset{}
+	if ctx.ShouldBind(&passwordReset) != nil {
+		ctx.JSON(400, gin.H{"error": "Invalid password reset request"})
+		return
+	}
+
+	user := ctx.GetString("User")
```

Comment:
The original code had a check for `if user == ""`. Was that removed because it was made redundant elsewhere in the code or should that check be added back in?

Reply:
Yes. Now `authHandler` checks if user is present and abort&return 401 if not.